javascript/ql/src/Declarations/UnusedProperty.ql

```diff
@@ -0,0 +1,59 @@
+/**
+ * @name Unused property
+ * @description Unused properties may be a symptom of a bug and should be examined carefully.
+ * @kind problem
+ * @problem.severity recommendation
+ * @id js/unused-property
+ * @tags maintainability
+ * @precision high
+ */
+
+import javascript
+import semmle.javascript.dataflow.CapturedNodes
+import UnusedVariable
+import UnusedParameter
+
+predicate hasUnknownPropertyRead(CapturedSource obj) {
+  // dynamic reads
+  exists(DataFlow::PropRead r | obj.getAPropertyRead() = r | not exists(r.getPropertyName()))
+  or
+  // reflective reads
+  obj.flowsToExpr(any(EnhancedForLoop l).getIterationDomain())
+  or
+  obj.flowsToExpr(any(InExpr l).getRightOperand())
+  or
+  obj.flowsToExpr(any(SpreadElement e).getOperand())
+  or
+  exists(obj.getAPropertyRead("hasOwnProperty"))
+  or
+  exists(obj.getAPropertyRead("propertyIsEnumerable"))
+  or
+  // string coercions
+  obj.flowsToExpr(any(AddExpr e).getAnOperand())
+  or
+  exists (obj.getAMethodCall("toString"))
+  or
+  exists (obj.getAMethodCall("valueOf"))
+}
+
+from DataFlow::PropWrite w, CapturedSource n, string name
+where
+  w = n.getAPropertyWrite(name) and
+  not exists(n.getAPropertyRead(name)) and
+  not w.getBase().analyze().getAValue() != n.analyze().getAValue() and
+  not hasUnknownPropertyRead(n) and
+  // avoid reporting if the definition is unreachable
+  w.getAstNode().getFirstControlFlowNode().getBasicBlock() instanceof ReachableBasicBlock and
+  // avoid implicitly read properties
+  not (
+    name = "toString" or
+    name = "valueOf" or
+    name.matches("@@%") // @@iterator, for example
+  ) and
+  // flagged by js/unused-local-variable
+  not exists(UnusedLocal l | l.getAnAssignedExpr().getUnderlyingValue().flow() = n) and
+  // flagged by js/unused-parameter
+  not exists(Parameter p | isAnAccidentallyUnusedParameter(p) |
+    p.getDefault().getUnderlyingValue().flow() = n
+  )
+select w, "Unused property " + name + "."
```

Comment:
In a typed JS dialect, the object may be stored in a local variable that has a type annotation, or be used in a type cast. That could lead to a compilation error if the property is removed. Could you whitelist a few cases like that?